`azurerm_managed_disk` - support expanding disk size without downtime for all types of `storage_account_type`

ms-zhenhua opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

According to [MS-DOC](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/expand-disks?tabs=azure-cli%2Cubuntu#expand-without-downtime), `azurerm_managed_disk` should support expanding without downtime for all types of `storage_account_type`

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [ ] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

--- PASS: TestAccManagedDisk_attachedDiskUpdate (624.88s)
--- PASS: TestAccManagedDisk_attachedDiskUpdateWithDowntime (567.79s)
--- PASS: TestAccManagedDisk_attachedNvmeDiskUpdateWithDowntime (565.66s)
--- PASS: TestAccManagedDisk_attachedPremiumV2DiskUpdateWithoutDowntime (451.35s)
--- PASS: TestAccManagedDisk_attachedUltraDiskUpdateWithoutDowntime (469.52s)

--- PASS: TestAccManagedDisk_fromPlatformImage (187.13s)
--- PASS: TestAccManagedDisk_empty (197.46s)
--- PASS: TestAccManagedDisk_networkPolicy (200.78s)
--- PASS: TestAccManagedDisk_importEmpty_withZone (221.15s)
--- PASS: TestAccManagedDisk_premiumV2WithIOpsReadOnlyAndMBpsReadOnly (279.64s)
--- PASS: TestAccManagedDisk_copy (152.12s)
--- PASS: TestAccManagedDisk_optimizedFrequentAttach (156.60s)
--- PASS: TestAccManagedDisk_performancePlus (126.39s)
--- PASS: TestAccManagedDisk_import (319.17s)
--- PASS: TestAccManagedDisk_update (163.14s)
--- PASS: TestAccManagedDisk_diskEncryptionSet_update (618.17s)
--- PASS: TestAccManagedDisk_upload (169.91s)
--- PASS: TestAccManagedDisk_attachedStorageTypeUpdate (690.34s)
--- PASS: TestAccManagedDisk_attachedTierUpdate (702.50s)
--- PASS: TestAccManagedDisk_premiumV2WithIOpsReadWriteAndMBpsReadWrite (221.42s)
--- PASS: TestAccManagedDisk_storageAccountType (167.86s)
--- PASS: TestAccManagedDisk_create_withHyperVGeneration (171.28s)
--- PASS: TestAccManagedDisk_create_withOnDemandBurstingEnabled (178.29s)
--- PASS: TestAccManagedDisk_import_withUltraSSD (153.47s)
--- PASS: TestAccManagedDisk_encryption (621.32s)
--- PASS: TestAccManagedDisk_update_withIOpsReadOnlyAndMBpsReadOnly (211.46s)
--- PASS: TestAccManagedDisk_update_withOnDemandBurstingEnabled (243.16s)
--- PASS: TestAccManagedDisk_create_withSecureVMDiskEncryptionSetId (475.66s)
--- PASS: TestAccManagedDisk_update_withUltraSSD (159.25s)
--- PASS: TestAccManagedDisk_zeroGbFromPlatformImage (134.83s)
--- PASS: TestAccManagedDisk_create_withUltraSSD (169.41s)
--- PASS: TestAccManagedDisk_create_withTrustedLaunchEnabled (181.28s)
--- PASS: TestAccManagedDisk_create_withSecurityType (193.10s)
--- PASS: TestAccManagedDisk_detachFromVM (584.27s)
--- PASS: TestAccManagedDisk_create_withLogicalSectorSize (172.30s)
--- PASS: TestAccManagedDisk_publicNetworkAccessUpdated (238.62s)
--- PASS: TestAccManagedDisk_fromGalleryImage (1153.23s)
--- PASS: TestAccManagedDisk_update_withMaxShares (208.70s)
--- PASS: TestAccManagedDisk_publicNetworkAccessDisabled (173.43s)
--- PASS: TestAccManagedDisk_diskEncryptionSet (465.86s)
--- PASS: TestAccManagedDisk_requiresImport (159.53s)
--- PASS: TestAccManagedDisk_publicNetworkAccessDefault (176.43s)
--- PASS: TestAccManagedDisk_networkPolicy_update_withAllowPrivate (287.41s)
--- PASS: TestAccManagedDisk_networkPolicy_create_withAllowPrivate (211.65s)
--- PASS: TestAccManagedDisk_onlineLiveResize (728.97s)
--- PASS: TestAccManagedDisk_networkPolicy_update (232.47s)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_managed_disk` - support expanding disk size without downtime for all types of `storage_account_type`


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Fixes [#28603](https://github.com/hashicorp/terraform-provider-azurerm/issues/28603)
Fixes [#26550](https://github.com/hashicorp/terraform-provider-azurerm/issues/26550)

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

stephybun reviewed (changes requested):
Thanks @ms-zhenhua, in addition to the comments I think the documentation might also want to be reviewed and updated to reflect these changes.

ms-zhenhua commented on internal/services/compute/no_downtime_resize.go:
```diff
@@ -24,56 +24,49 @@ import (
 		return pointer.To(false), nil
 	}
 
+	virtualMachineIdRaw := disk.ManagedBy
```
unchanged. Because `virtualMachineIdRaw` is also used by `log.Printf` in line 76.

ms-zhenhua commented on internal/services/compute/no_downtime_resize.go:
```diff
@@ -24,56 +24,49 @@ import (
 	}
 
+	virtualMachineIdRaw := disk.ManagedBy
 	virtualMachineId, err := virtualmachines.ParseVirtualMachineIDInsensitively(*virtualMachineIdRaw)
```
unchanged. The reason is https://github.com/hashicorp/terraform-provider-azurerm/pull/28730/files#r2002677665

ms-zhenhua commented on internal/services/compute/no_downtime_resize.go:
```diff
@@ -84,12 +77,27 @@ func determineIfVirtualMachineSkuSupportsNoDowntimeResize(ctx context.Context, v
+			if model.Properties.StorageProfile != nil && model.Properties.StorageProfile.DiskControllerType != nil {
+				vmDiskControllerType = string(*model.Properties.StorageProfile.DiskControllerType)
+			}
 		}
```
Updated. Thanks.

ms-zhenhua commented on internal/services/compute/no_downtime_resize.go:
```diff
@@ -24,56 +24,49 @@ import (
-	if oldSizeGb < 4096 && newSizeGb >= 4096 {
-		return pointer.To(false)
+	isPremiumV2DiskOrUltra := strings.EqualFold(string(*disk.Sku.Name), string(disks.DiskStorageAccountTypesPremiumVTwoLRS)) || strings.EqualFold(string(*disk.Sku.Name), string(disks.DiskStorageAccountTypesUltraSSDLRS))
+	if !isPremiumV2DiskOrUltra && oldSizeGb < 4096 && newSizeGb >= 4096 {
```
updated. Thanks.

ms-zhenhua commented on internal/services/compute/no_downtime_resize.go:
```diff
@@ -104,14 +112,9 @@ func determineIfVirtualMachineSkuSupportsNoDowntimeResize(ctx context.Context, v
-			continue
-		}
-		if sku.Capabilities == nil {
+		if !strings.EqualFold(pointer.From(sku.ResourceType), "virtualMachines") || sku.Capabilities == nil || !strings.EqualFold(pointer.From(sku.Name), vmSku) {
```
updated. Thanks.

ms-zhenhua commented on internal/services/compute/managed_disk_resource.go:
```diff
@@ -725,17 +725,16 @@ func resourceManagedDiskUpdate(d *pluginsdk.ResourceData, meta interface{}) erro
 			canBeResizedWithoutDowntime := false
 			if meta.(*clients.Client).Features.ManagedDisk.ExpandWithoutDowntime {
-				diskSupportsNoDowntimeResize := determineIfDataDiskSupportsNoDowntimeResize(disk.Model, oldSize.(int), newSize.(int))
+				diskSupportsNoDowntimeResize, needToDetach := determineIfDataDiskSupportsNoDowntimeResize(disk.Model, oldSize.(int), newSize.(int))
```
updated. I added a new method `determineIfDataDiskRequiresDetaching` and make the return value as a input parameter of `determineIfDataDiskSupportsNoDowntimeResize`.

ms-zhenhua commented:
> Thanks @ms-zhenhua, in addition to the comments I think the documentation might also want to be reviewed and updated to reflect these changes.

Hi @stephybun , 
Thank you for the review. I did not update the [documentation](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/managed_disk#disk_size_gb-1) because I found in the original document, the `Note` of No Downtime Resizing directly points to the MS-DOC so that it seems we do not need to update the Terraform document. Besides, I have updated the code and comments. Could you please help take another review? Thanks.

stephybun commented:
@ms-zhenhua the no downtime resizing note in the docs is fine, but the note on when a disk will be detached isn't correct anymore: 
https://github.com/hashicorp/terraform-provider-azurerm/blob/244f165be2fdb42eccbcbd4bfe7f947382a5ef3f/website/docs/r/managed_disk.html.markdown?plain=1#L111
Can you please update that note?

ms-zhenhua commented:
> @ms-zhenhua the no downtime resizing note in the docs is fine, but the note on when a disk will be detached isn't correct anymore:
> 
> https://github.com/hashicorp/terraform-provider-azurerm/blob/244f165be2fdb42eccbcbd4bfe7f947382a5ef3f/website/docs/r/managed_disk.html.markdown?plain=1#L111
> 
> 
> Can you please update that note?

Hi @stephybun, thanks for the explanation. I have updated the document. Please help take another review.

stephybun reviewed (approved):
Thanks @ms-zhenhua LGTM 👍
